persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/ClinicalDataSparkRepository.java

```diff
@@ -0,0 +1,280 @@
+
+                String query = sb.toString();
+                Dataset<Row> sub = spark.sql(query);
+
+                sub = sub.withColumn("attrId", lit(c))
+                    .withColumn("studyId", lit(studyId));
+
+                if (MUTATION_COUNT.equalsIgnoreCase(c)) {
+                    sub = sub.join(getMutationCount(studyId), "sampleId");
+                }
+                if (FRACTION_GENOME_ALTERED.equalsIgnoreCase(c)) {
+                    sub = sub.join(getFractionGenomeAltered(studyId), "sampleId");
+                }
+                res.add(sub);
+            }
+        }
+        Dataset<Row> ds = res.get(0);
+        for (Dataset<Row> r : res.subList(1, res.size())) {
+            ds = ds.unionByName(r);
+        }
+
+        Dataset<ClinicalDataModel> clinicalDataModel = ds.as(Encoders.bean(ClinicalDataModel.class));
+        List<ClinicalDataModel> dsls = clinicalDataModel.collectAsList();
+        List<ClinicalData> clinicalData = dsls.stream()
+            .map(r -> mapToClinicalData(r)).collect(Collectors.toList());
+        return clinicalData;
+    }
+
+    @Override
+    public BaseMeta fetchMetaClinicalData(List<String> studyIds, List<String> ids, List<String> attributeIds, String clinicalDataType) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public List<ClinicalDataCount> fetchClinicalDataCounts(List<String> studyIds, List<String> sampleIds,
+                                                           List<String> attributeIds, String clinicalDataType) {
+        List<Dataset<Row>> res = new ArrayList<>();
+
+        spark.sqlContext().setConf("spark.sql.caseSensitive", "false");
+        Set<String> studyIdSet = new HashSet<>(studyIds);
+        String dataFile = ParquetConstants.CLINICAL_PATIENT;
+        if (PersistenceConstants.SAMPLE_CLINICAL_DATA_TYPE.equalsIgnoreCase(clinicalDataType)) {
+            dataFile = ParquetConstants.CLINICAL_SAMPLE;
+        }
+        Dataset<Row> clinicalData = loadDataFiles(studyIdSet, "/" + dataFile);
+
+        if ("PATIENT".equalsIgnoreCase(clinicalDataType)) {
+            Dataset<Row> clinicalSampleData = loadDataFiles(studyIdSet, ParquetConstants.CLINICAL_SAMPLE);
+            clinicalData = clinicalData.join(clinicalSampleData, "PATIENT_ID");
+        }
+
+        // filter by sample id
+        if (sampleIds != null && !sampleIds.isEmpty()) {
+            String[] sampleArr = sampleIds.stream().toArray(String[]::new);
+            clinicalData = clinicalData.filter(col("SAMPLE_ID").isin(sampleArr));
+        }
+
+        // if attributes is null return data from all columns
+        List<String> attributes = attributeIds;
+        if (attributeIds == null || attributeIds.isEmpty()) {
+            attributes = Arrays.asList(clinicalData.columns());
+        }
+
+        clinicalData.createOrReplaceTempView("clinicalData");
+        for (String c : attributes) {
+            StringBuilder sb = new StringBuilder();
```

Comment:
not sure if there is a better way to do instead of iterating for each attribute. response time increases as the number of attributes increases

Reply:
yeah, I'm not sure as i'm using attributeId for the value and group by.

Reply:
I tried adding union as part of the sql, it does not show a big impact on performance. I'm not sure if there is a way to do multiple group bys in the same select...![clinical-data-bin-counts](https://user-images.githubusercontent.com/2739448/61876051-f12dbc80-aeb9-11e9-90e9-5432ae913a69.png)

Reply:
Leave it as is for now until we find a better solution.

Reply:
ok